File: ios/ExpoVlcPlayerModule.swift
import ExpoModulesCore

public final class ExpoVlcPlayerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoVlcPlayer")

    View(ExpoVlcPlayerView.self) {
      Prop("url") { (view: ExpoVlcPlayerView, value: String?) in
        DispatchQueue.main.async {
          view.setStreamUrl(value)
        }
      }

      Prop("paused") { (view: ExpoVlcPlayerView, paused: Bool) in
        DispatchQueue.main.async {
          view.setPaused(paused)
        }
      }

      Prop("initOptions") { (view: ExpoVlcPlayerView, options: [String]?) in
        DispatchQueue.main.async {
          view.setInitOptions(options)
        }
      }

      Prop("mediaOptions") { (view: ExpoVlcPlayerView, options: [String]?) in
        DispatchQueue.main.async {
          view.setMediaOptions(options)
        }
      }

      Prop("videoAspectRatio") { (view: ExpoVlcPlayerView, ratio: String?) in
        DispatchQueue.main.async {
          view.setAspectRatio(ratio)
        }
      }

      Prop("resizeMode") { (view: ExpoVlcPlayerView, mode: String?) in
        DispatchQueue.main.async {
          view.setResizeMode(mode)
        }
      }

      Events("onLoad", "onPlaying", "onError")
    }

    AsyncFunction("retry") { (viewTag: Int) -> Void in
      guard let appContext = self.appContext,
            let view = appContext.findView(withTag: viewTag, ofType: ExpoVlcPlayerView.self) else {
        throw Exceptions.ViewNotFound((tag: viewTag, type: ExpoVlcPlayerView.self))
      }
      view.retry()
    }
  }
}

File: android/src/main/java/expo/modules/vlcplayer/ExpoVlcPlayerModule.kt
package expo.modules.vlcplayer

import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoVlcPlayerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoVlcPlayer")

    View(ExpoVlcPlayerView::class) {
      Prop("url") { view: ExpoVlcPlayerView, url: String? ->
        view.setStreamUrl(url)
      }

      Prop("paused") { view: ExpoVlcPlayerView, paused: Boolean ->
        if (paused) {
          view.pause()
        } else {
          view.resume()
        }
      }

      Prop("initOptions") { view: ExpoVlcPlayerView, options: List<String>? ->
        view.setInitOptions(options)
      }

      Prop("mediaOptions") { view: ExpoVlcPlayerView, options: List<String>? ->
        view.setMediaOptions(options)
      }

      Prop("videoAspectRatio") { view: ExpoVlcPlayerView, aspectRatio: String? ->
        view.setAspectRatio(aspectRatio)
      }

      Prop("resizeMode") { view: ExpoVlcPlayerView, resizeMode: String? ->
        view.setResizeMode(resizeMode)
      }

      Events("onLoad", "onPlaying", "onError")

      OnViewDestroys { view: ExpoVlcPlayerView ->
        view.release()
      }
    }

    AsyncFunction("retry") { viewTag: Int ->
      val view = appContext.findView<ExpoVlcPlayerView>(viewTag)
        ?: throw Exceptions.ViewNotFound(ExpoVlcPlayerView::class, viewTag)

      view.retry()
    }
  }
}
